sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs

```diff
@@ -65,104 +255,393 @@ public void AddLink(string id, IDictionary<string, string>? attributes = null)
                     }
                 }
 
-                _activity.AddLink(linkedActivity);
+                _links ??= new List<Activity>();
+                _links.Add(linkedActivity);
             }
-        }
 
-        public void Start()
-        {
-            if (_activity != null)
+            public void Start()
             {
-                _source.StartActivity(_activity, _activity);
+                _currentActivity = StartActivitySourceActivity();
+
+                if (_currentActivity == null)
+                {
+                    if (!_diagnosticSource.IsEnabled(_activityName, _diagnosticSourceArgs))
+                    {
+                        return;
+                    }
+
+                    _currentActivity = new DiagnosticActivity(_activityName)
+                    {
+                        Links = (IEnumerable<Activity>?)_links ?? Array.Empty<Activity>(),
+                    };
+                    _currentActivity.SetW3CFormat();
+                    _currentActivity.SetStartTime(_startTime.DateTime);
+
+                    if (_tagCollection != null)
+                    {
+                        foreach (var tag in _tagCollection)
+                        {
+                            _currentActivity.AddTag(tag.Key, (string)tag.Value);
+                        }
+                    }
+
+                    _currentActivity.Start();
+                }
+
+                _diagnosticSource.Write(_activityName + ".Start", _diagnosticSourceArgs ?? _currentActivity);
+            }
+
+            private Activity? StartActivitySourceActivity()
+            {
+                return ActivityExtensions.ActivitySourceStartActivity(
+                    _activitySource,
+                    _activityName,
+                    (int)_kind,
+                    startTime: _startTime,
+                    tags: _tagCollection,
+                    links: GetActivitySourceLinkCollection());
+            }
+
+            public void SetStartTime(DateTime startTime)
+            {
+                _startTime = startTime;
+                _currentActivity?.SetStartTime(startTime);
+            }
+
+            public void MarkFailed(Exception exception)
+            {
+                _diagnosticSource?.Write(_activityName + ".Exception", exception);
+            }
+
+            public void Dispose()
+            {
+                if (_currentActivity == null)
+                {
+                    return;
+                }
+
+                _diagnosticSource.Write(_activityName + ".Stop", _diagnosticSourceArgs);
+
+                if (!_currentActivity.TryDispose())
```

Comment:
Having seen the implementation, how would `Stop` work if you've already disposed it like here? If it does, I find that a little alarming since dispose was supposed to tear down any resources that needed special tear down, so I would think / hope `Stop` would fail. That's typical in most (all?) BCL classes that implement `IDisposable`.

Reply:
`_currentActivity.TryDispose` will return false if the current version of the DiagnosticSource library doesn't have the Dispose method. In that case, we fall back to .Stop